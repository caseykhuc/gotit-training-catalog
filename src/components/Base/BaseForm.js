import React from 'react';
import { Form, Alert } from 'react-bootstrap';
import PropTypes from 'prop-types';

const BaseForm = ({
  onInputChange,
  inputError,
  requestError,
  inputValue,
  fields,
  onKeyDown,
}) => (
    <div onKeyDown={onKeyDown}>
      {fields.map(({ name, type, options }) => (
        <Form.Group controlId={name} key={name}>
          {type !== 'select'
            ? (
              <Form.Control
                type={type}
                name={name}
                value={inputValue[name]}
                placeholder={`Enter ${name}`}
                onChange={onInputChange}
                isInvalid={inputError[name]}
              />
            )
            : (
              <Form.Control
                as="select"
                name={name}
                onChange={onInputChange}
                isInvalid={inputError[name]}
              >
                {options.map(({ name, id }) => <option value={id} key={id}>{`Category ${name}`}</option>)}
              </Form.Control>
            )}
          <Form.Control.Feedback type="invalid">{inputError[name]}</Form.Control.Feedback>
        </Form.Group>
      ))}
      {requestError && <Alert variant="danger">{requestError}</Alert>}
    </div>
  )

BaseForm.propTypes = {
  onInputChange: PropTypes.func.isRequired,
  onKeyDown: PropTypes.func.isRequired,
  inputError: PropTypes.object.isRequired,
  inputValue: PropTypes.object.isRequired,
  fields: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    value: PropTypes.string,
  })),
  requestError: PropTypes.string,
}

export default BaseForm;
