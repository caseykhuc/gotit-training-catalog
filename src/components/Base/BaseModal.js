import React from 'react';
import { Modal, Button } from 'react-bootstrap';
import { connect } from 'react-redux';
import PropTypes from 'prop-types'
import { hideModal } from '../../actions/modal';

export const BaseModal = ({
  title, hideModal, onAccept, children,
}) => (
    <Modal show onHide={hideModal} centered>
      <Modal.Header closeButton>
        <Modal.Title>{title}</Modal.Title>
      </Modal.Header>
      <Modal.Body>{children}</Modal.Body>
      <Modal.Footer>
        <Button variant="secondary" onClick={hideModal}>
          Close
      </Button>
        <Button variant="primary" onClick={onAccept}>
          Accept
      </Button>
      </Modal.Footer>
    </Modal>
  )

BaseModal.propTypes = {
  title: PropTypes.string.isRequired,
  hideModal: PropTypes.func.isRequired,
  onAccept: PropTypes.func.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
}

export default connect(null, { hideModal })(BaseModal);
