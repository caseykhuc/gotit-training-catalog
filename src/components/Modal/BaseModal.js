import React from 'react';
import { Modal, Button } from 'react-bootstrap';
import { hideModal } from '../../actions';
import { connect } from 'react-redux';

class BaseModal extends React.Component {
  render() {
    return (
      <Modal.Dialog>
        <Modal.Header closeButton onClick={this.props.hideModal}>
          <Modal.Title>Modal title</Modal.Title>
        </Modal.Header>

        <Modal.Body>
          <p>Modal body text goes here.</p>
        </Modal.Body>

        <Modal.Footer>
          <Button variant='secondary'>Close</Button>
          <Button variant='primary'>Save changes</Button>
        </Modal.Footer>
      </Modal.Dialog>
    );
  }
}

export default connect(null, { hideModal })(BaseModal);
