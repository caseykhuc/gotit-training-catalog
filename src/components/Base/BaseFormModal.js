import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import BaseModal from './BaseModal';
import BaseForm from './BaseForm';

class BaseFormModal extends React.Component {
  state = this.props.initialState

  onInputChange = (e) => {
    const { name, value } = e.target;
    const { inputValue } = this.state;

    const newValue = { ...inputValue, [name]: value };
    const { validate } = this.props;

    this.setState(validate ? {
      inputValue: newValue,
      inputError: validate(newValue),
    } : {
        inputValue: newValue,
      });
  }

  onFormSubmit = async () => {
    const { onAction, initialState, validate } = this.props;
    const { inputValue, inputError } = this.state;

    console.log(inputValue);
    if (!_.values(inputValue).some((x) => !_.isEmpty(x))) {
      this.setState({ requestError: 'Empty input' });
      return;
    }

    if (validate && !_.isEmpty(validate(inputValue))) {
      return;
    }

    if (inputError) {
      const res = await onAction(inputValue);
      if (!res.success) {
        this.setState((typeof res.message === 'object')
          ? { inputError: res.message }
          : { inputError: {}, requestError: res.message })
      } else this.setState(initialState)
    }
  }

  onKeyDown = (e) => {
    if (e.key === 'Enter') {
      this.onFormSubmit();
    }
  }

  render() {
    const {
      inputValue, inputError, requestError,
    } = this.state;

    const { fields, title } = this.props;
    return (
      <BaseModal title={title} onAccept={() => this.onFormSubmit()}>
        <BaseForm
          onInputChange={(e) => this.onInputChange(e)}
          inputValue={inputValue}
          inputError={inputError}
          requestError={requestError}
          fields={fields}
          onKeyDown={(e) => this.onKeyDown(e)}
        />
      </BaseModal>
    )
  }
}

BaseFormModal.propTypes = {
  fields: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    value: PropTypes.string,
  })),
  initialState: PropTypes.shape({
    inputValue: PropTypes.object,
    inputError: PropTypes.object,
    requestError: PropTypes.string,
  }),
  onAction: PropTypes.func.isRequired,
  title: PropTypes.string,
  validate: PropTypes.func,
}

export default BaseFormModal;
