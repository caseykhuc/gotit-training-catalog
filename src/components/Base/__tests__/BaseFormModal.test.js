import React from 'react';
import { shallow } from 'enzyme';
import BaseForm from '../BaseForm'
import BaseModal from '../BaseModal';
import BaseFormModal from '../BaseFormModal';

describe('components/BaseFormModal', () => {
  let props;
  let wrapper;
  let baseModal;
  let baseForm;

  const update = () => {
    wrapper.update();
    baseModal = wrapper.find(BaseModal);
    baseForm = wrapper.find(BaseForm);
  };
  const setup = () => {
    wrapper = shallow(<BaseFormModal {...props} />);
    update();
  };
  beforeEach(() => {
    props = {
      onAction: jest.fn(),
      initialState: {
        inputError: {},
        inputValue: {
          username: 'test',
          password: 'test',
        },
        requestError: 'error test',
      },
      fields: [{ name: 'username', type: 'text' }, {
        name: 'password', type: 'password',
      }],
      title: 'Test Title',
      /* validate: () => { }, */
    }
  });
  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should invoke onAction when user press Enter', () => {
    setup();
    baseForm.props().onKeyDown({ key: 'Enter' });
    expect(props.onAction).toHaveBeenCalled();
  });
  it('should not invoke onAction when user press other key', () => {
    setup();
    baseForm.props().onKeyDown({ key: 'b' });
    expect(props.onAction).not.toHaveBeenCalled();
  });
  it('should invoke onAction when baseFormModal invoked onAccept ', () => {
    setup();
    baseModal.props().onAccept();
    expect(props.onAction).toHaveBeenCalled();
  })
  it('should controll inputs in BaseForm', () => {
    setup();
    baseForm.props().onInputChange({ target: { name: 'username', value: 'changed' } });
    expect(wrapper.state().inputValue.username).toEqual('changed');
  })
  it('should set request error coresponding to failure message', async () => {
    props.onAction = () => Promise.resolve({ success: false, message: 'error test changed' });
    setup();
    await baseModal.props().onAccept();
    expect(wrapper.state().requestError).toEqual('error test changed')

    props.onAction = () => Promise.resolve({ success: false, message: { username: 'test changed' } });
    setup();
    await baseModal.props().onAccept();
    expect(wrapper.state().inputError).toEqual({ username: 'test changed' });
  })
  it('should reset state value when receive a success response', async () => {
    props.onAction = () => Promise.resolve({ success: true });
    setup();
    await baseModal.props().onAccept();
    expect(wrapper.state()).toEqual(props.initialState);
  })
})
