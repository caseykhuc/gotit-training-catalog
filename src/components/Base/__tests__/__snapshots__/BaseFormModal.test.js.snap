// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/BaseFormModal should render correctly 1`] = `ShallowWrapper {}`;
